lib/galaxy/objectstore/irods.py

```diff
@@ -0,0 +1,710 @@
+        #     # print("***1 %s exists" % cache_path)
+        #     if self._key_exists(rel_path):
+        #         # print("***2 %s exists in S3" % rel_path)
+        #         # Make sure the size in cache is available in its entirety
+        #         # print("File '%s' cache size: %s, S3 size: %s" % (cache_path, os.path.getsize(cache_path), self._get_size_in_s3(rel_path)))
+        #         if os.path.getsize(cache_path) == self._get_size_in_s3(rel_path):
+        #             # print("***2.1 %s exists in S3 and the size is the same as in cache (in_cache=True)" % rel_path)
+        #             exists = True
+        #         else:
+        #             # print("***2.2 %s exists but differs in size from cache (in_cache=False)" % cache_path)
+        #             exists = False
+        #     else:
+        #         # Although not perfect decision making, this most likely means
+        #         # that the file is currently being uploaded
+        #         # print("***3 %s found in cache but not in S3 (in_cache=True)" % cache_path)
+        #         exists = True
+        # else:
+
+    def _pull_into_cache(self, rel_path):
+        # Ensure the cache directory structure exists (e.g., dataset_#_files/)
+        rel_path_dir = os.path.dirname(rel_path)
+        if not os.path.exists(self._get_cache_path(rel_path_dir)):
+            os.makedirs(self._get_cache_path(rel_path_dir))
+        # Now pull in the file
+        file_ok = self._download(rel_path)
+        self._fix_permissions(self._get_cache_path(rel_path_dir))
+        return file_ok
+
+    def _download(self, rel_path):
+        log.debug("Pulling data object '%s' into cache to %s", rel_path, self._get_cache_path(rel_path))
+
+        p = Path(rel_path)
+        data_object_name =  p.stem + p.suffix
+        subcollection_name = p.parent
+
+        collection_path = self.home + "/" + str(subcollection_name)
+        data_object_path = collection_path + "/" + str(data_object_name) 
+        data_obj = None
+
+        try:
+            data_obj = self.session.data_objects.get(data_object_path)
+        except (DataObjectDoesNotExist, CollectionDoesNotExist):
+            log.warn("Collection or data object (%s) does not exist", data_object_path)
+            return False 
+
+        if self.cache_size > 0 and data_obj.__sizeof__() > self.cache_size:
+            log.critical("File %s is larger (%s) than the cache size (%s). Cannot download.",
+                         rel_path, data_obj.__sizeof__(), self.cache_size)
+            return False
+
+        log.debug("Pulled data object '%s' into cache to %s", rel_path, self._get_cache_path(rel_path))
+        
+        with data_obj.open('rb') as data_obj_fp, open(self._get_cache_path(rel_path), "wb") as cache_fp:
+            content = data_obj_fp.read()
+            cache_fp.write(content)
+        return True
+
+    def _push_to_os(self, rel_path, source_file=None, from_string=None):
```

Comment:
Since this is pushing to irods maybe `push_to_irods` is a better name ?

Reply:
Sure